cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesDataset.java

```diff
@@ -0,0 +1,208 @@
+ * It does not wrap its operations in a transaction. It is up to the caller to decide what operations belong
+ * in a transaction.
+ */
+public class PreferencesDataset {
+  private static final String PREFERENCES_CONFIG_TYPE = "preferences";
+  // Namespace where instance level properties are stored
+  private static final String EMPTY_NAMESPACE = "";
+  // Id for Properties config stored at the instance level
+  private static final String INSTANCE_PROPERTIES = "instance";
+
+  private final ConfigDataset configDataset;
+
+  private PreferencesDataset(ConfigDataset configDataset) {
+    this.configDataset = configDataset;
+  }
+
+  public static PreferencesDataset get(DatasetContext datasetContext, DatasetFramework dsFramework) {
+    return new PreferencesDataset(ConfigDataset.get(datasetContext, dsFramework));
+  }
+
+  /**
+   * Get the preferences for the entity id
+   *
+   * @param entityId the entity id to get the preferences from
+   * @return the map which contains the preferences
+   */
+  public Map<String, String> getPreferences(EntityId entityId) {
+    switch (entityId.getEntityType()) {
+      case INSTANCE:
+        return getConfigProperties(EMPTY_NAMESPACE, getMultipartKey(INSTANCE_PROPERTIES));
+      case NAMESPACE:
+        NamespaceId namespaceId = (NamespaceId) entityId;
+        return getConfigProperties(namespaceId.getNamespace(), getMultipartKey(namespaceId.getNamespace()));
+      case APPLICATION:
+        ApplicationId appId = (ApplicationId) entityId;
+        return getConfigProperties(appId.getNamespace(), getMultipartKey(appId.getNamespace(), appId.getApplication()));
+      case PROGRAM:
+        ProgramId programId = (ProgramId) entityId;
+        return getConfigProperties(programId.getNamespace(),
+                                   getMultipartKey(programId.getNamespace(), programId.getApplication(),
+                                                   programId.getType().getCategoryName(), programId.getProgram()));
+      default:
+        throw new UnsupportedOperationException(
+          String.format("Preferences cannot be used on this entity type: %s", entityId.getEntityType()));
+    }
+  }
+
+  /**
+   * Get the resolved preferences for the entity id, the preferences are resolved from instance -> namespace -> app
+   * -> program level
+   *
+   * @param entityId the entity id to get the preferences from
+   * @return the map which contains the preferences
+   */
+  public Map<String, String> getResolvedPreferences(EntityId entityId) {
+    // if it is instance level get the properties and return
+    if (entityId.getEntityType().equals(EntityType.INSTANCE)) {
+      return getPreferences(entityId);
+    }
+
+    Map<String, String> properties = new HashMap<>();
+    // if the entity id has a parent id, get the preference from its parent
+    if (entityId instanceof ParentedId) {
+      properties.putAll(getResolvedPreferences(((ParentedId) entityId).getParent()));
+    } else {
```

Comment:
it's safer to make this `else if (entityId.getType() == EntityType.NAMESPACE)` in case EntityId changes in the future

Reply:
Fixed